sdk/iot/provisioning/samples/src/paho_iot_provisioning_example.c

```diff
@@ -0,0 +1,304 @@
+static int subscribe()
+{
+  int rc;
+
+  char topic_filter[128];
+  az_span topic_filter_span = AZ_SPAN_LITERAL_FROM_BUFFER(topic_filter);
+  if ((rc = az_iot_provisioning_client_register_subscribe_topic_filter_get(
+           &provisioning_client, topic_filter_span, &topic_filter_span))
+      != AZ_OK)
+
+  {
+    printf("Failed to get MQTT SUB topic filter, return code %d\n", rc);
+    return rc;
+  }
+
+  if ((rc = az_span_append_uint8(topic_filter_span, '\0', &topic_filter_span)) != AZ_OK)
+  {
+    printf("Failed to get MQTT SUB topic filter, return code %d\n", rc);
+    return rc;
+  }
+
+  if ((rc = MQTTClient_subscribe(mqtt_client, topic_filter, 1)) != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to subscribe, return code %d\n", rc);
+    return rc;
+  }
+
+  return 0;
+}
+
+static int register_device()
+{
+  int rc;
+  MQTTClient_message pubmsg = MQTTClient_message_initializer;
+
+  char topic[128];
+  az_span topic_span = AZ_SPAN_LITERAL_FROM_BUFFER(topic);
+
+  if ((rc = az_iot_provisioning_client_register_publish_topic_get(
+           &provisioning_client, topic_span, &topic_span))
+      != AZ_OK)
+
+  {
+    printf("Failed to get MQTT PUB register topic, return code %d\n", rc);
+    return rc;
+  }
+
+  if ((rc = az_span_append_uint8(topic_span, '\0', &topic_span)) != AZ_OK)
+  {
+    printf("Failed to get MQTT PUB register topic, return code %d\n", rc);
+    return rc;
+  }
+
+  pubmsg.payload = NULL; // empty payload.
+  pubmsg.payloadlen = 0;
+  pubmsg.qos = 1;
+  pubmsg.retained = 0;
+
+  if ((rc = MQTTClient_publishMessage(mqtt_client, topic, &pubmsg, NULL)) != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to publish register request, return code %d\n", rc);
+    return rc;
+  }
+
+  return 0;
+}
+
+int main()
+{
+  int rc;
+
+  if ((rc = read_configuration()) != AZ_OK)
+  {
+    printf("Failed to read configuration from environment variables, return code %d\n", rc);
+    return rc;
+  }
+
+  char client_id[128];
+  az_span client_id_span = AZ_SPAN_LITERAL_FROM_BUFFER(client_id);
+  if ((rc
+       = az_iot_provisioning_client_id_get(&provisioning_client, client_id_span, &client_id_span))
+      != AZ_OK)
+
+  {
+    printf("Failed to get MQTT clientId, return code %d\n", rc);
+    return rc;
+  }
+
+  if ((rc = az_span_append_uint8(client_id_span, '\0', &client_id_span)) != AZ_OK)
+  {
+    printf("Failed to get MQTT username, return code %d\n", rc);
+    return rc;
+  }
+
+  if ((rc = MQTTClient_create(
+           &mqtt_client,
+           global_provisioning_endpoint,
+           client_id,
+           MQTTCLIENT_PERSISTENCE_NONE,
+           NULL))
+      != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to create MQTT client, return code %d\n", rc);
+    return rc;
+  }
+
+  if ((rc = MQTTClient_setCallbacks(mqtt_client, NULL, NULL, on_received, NULL))
+      != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to set MQTT callbacks, return code %d\n", rc);
+    return rc;
+  }
+
+  if ((rc = connect()))
+  {
+    return rc;
```

Comment:
Please clean client once it has been created successfully in every exit path.

Now, Valgrind looks to be imp to be run on this and also integrated into pipeline.

Reply:
Abnormal termination will immediately stop the process. Only normal termination will clean-up. This is very common, it's also what the official Paho sample does.

Reply:
How are you going to justify the valgrind output?

---
In reply to: [404466337](https://github.com/Azure/azure-sdk-for-c/pull/547#discussion_r404466337) [](ancestors = 404466337)

Reply:
Clean shutdown is what I prefer.

---
In reply to: [404483628](https://github.com/Azure/azure-sdk-for-c/pull/547#discussion_r404483628) [](ancestors = 404483628,404466337)